src/compiler/checker.ts

```diff
@@ -6730,6 +6729,7 @@ namespace ts {
                 // will never be observed because a qualified name can't reference signatures.
                 if (symbol.flags & (SymbolFlags.Function | SymbolFlags.Method)) {
                     type.callSignatures = getSignaturesOfSymbol(symbol);
+                    type.constructSignatures = filter(type.callSignatures, sig => isJSConstructor(sig.declaration));

```

Comment:
The flags on this raise an interesting question:

```ts
const obj = {
  /** @constructor */
  Foo() { this.bat = "bar" }
}
(new obj.Foo()).bar
```
Should actually work, right? Do we test this?

Reply:
well, uh, it's an error at runtime because obj.Foo isn't a constructor.

It turns out the reason it doesn't work in the checker is that `isJSConstructor` requires the tag to be on functions, not methods. I added a test, with noImplicitAny on so that the errors are visible.